### PR TITLE
Added Mujoco Env for Quanser Qube!

### PR DESCRIPTION
**Pull request by fdamken**
Furuta Pendulum

### CODE REVIEW

File: Pyrado/pyrado/environments/mujoco/quanser_qube.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        # Define the task including the reward function
+        state_des = task_args.get("state_des", np.array([0.0, np.pi, 0.0, 0.0]))
+        Q = task_args.get("Q", np.diag([1.0, 1.0, 2e-2, 5e-3]))  # former: [3e-1, 1.0, 2e-2, 5e-3]
+        R = task_args.get("R", np.diag([4e-3]))
+
+        return RadiallySymmDesStateTask(self.spec, state_des, ExpQuadrErrRewFcn(Q, R), idcs=[1])
+
+
+class QQubeStabMjSim(QQubeMjSim):
+    name: str = "qq-st-mj"
```
Comment (famura) [inline Pyrado/pyrado/environments/mujoco/quanser_qube.py:212]: same (qq-mj-st)

File: Pyrado/tests/conftest.py [hunk 2]
```diff
@@ -192,6 +193,16 @@ def default_qqst():
...
+    @staticmethod
+    @m_needs_mujoco
+    def default_qqst_mj():
+        return QQubeStabMjSim(dt=0.01, max_steps=500)
+
+    @staticmethod
+    @m_needs_mujoco
+    def default_qqsu_mj():
+        return QQubeSwingUpMjSim(dt=0.004, max_steps=4000)
+
```
Comment (famura) [inline Pyrado/tests/conftest.py:205]: good that you added them. name can be kept here.

File: Pyrado/tests/test_environments.py [hunk 3]
```diff
@@ -87,6 +89,8 @@
         "qq-su",
         "qcp-st",
         "qcp-su",
+        "qq-st-mj",
```
Comment (famura) [inline Pyrado/tests/test_environments.py:92]: you added the new env to the most relevant tests :+1:  please also add it to the test_setting_dp_vals() test, and locally check if it works

**Review by famura — CHANGES_REQUESTED** [review]

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        <texture name="floor" type="2d" file="IAS_floor_fine.png"/>
+        <texture name="qube" file="qqube.png" gridsize="3 4" gridlayout=".U..BFRL.D.."/>
+        <material name="floor" rgba="0.8 0.9 0.8 1" reflectance="0.1" texture="floor" texrepeat="10 10"/>
+        <material name="qube" reflectance="0.7" texture="qube"/>
+        <material name="link" rgba="0.75 0.75 0.75 1" reflectance="0.9"/>
+        <material name="pole" rgba="1 0 0 1" reflectance="0.9"/>
+    </asset>
+    <worldbody>
+        <light cutoff="100" diffuse="1 1 1" dir="-0 0 -1.3" directional="true" exponent="1" pos="0 0 1.3" specular=".1 .1 .1"/>
+        <geom name="ground" type="plane" pos="0 0 0" size="10 10 0.1" material="floor"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:55]: discussed in discord: it's not 10cm thick

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        <geom name="cube" type="box" size="0.05 0.06 0.05" pos="0 0 0.06" euler="90 90 0" material="qube"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:56]: I'll make it cubic, this was to not cause (purely visual) collisions of the pole with the cube or the floor.

[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:56]: for the texture to be correctly rotated

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        <geom name="upLink" type="cylinder" fromto="0 0 0.12 0 0 0.13" size="0.003" material="link" mass="0"/>
+        <geom name="mount" type="cylinder" fromto="0 0 0.12 0 0 0.1225" size="0.0075" material="qube" mass="0"/>
+        <body>
+            <joint name="theta" type="hinge" pos="0 0 0.13" axis="0 0 1" damping="[Dr]" frictionloss="0"/>
+            <geom name="rotaryArm" type="cylinder" fromto="0 0 0.13 [Lr] 0 0.13" size="0.003" material="link" mass="[Mr]"/>
+            <geom name="rotaryArmExtension" type="cylinder" fromto="[Lr] 0 0.13 [0.0055+Lr] 0 0.13" size="0.003" material="link" mass="0"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:62]: visual beauty only

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        <material name="pole" rgba="1 0 0 1" reflectance="0.9"/>
+    </asset>
+    <worldbody>
+        <light cutoff="100" diffuse="1 1 1" dir="-0 0 -1.3" directional="true" exponent="1" pos="0 0 1.3" specular=".1 .1 .1"/>
+        <geom name="ground" type="plane" pos="0 0 0" size="10 10 0.1" material="floor"/>
+        <geom name="cube" type="box" size="0.05 0.06 0.05" pos="0 0 0.06" euler="90 90 0" material="qube"/>
+        <geom name="upLink" type="cylinder" fromto="0 0 0.12 0 0 0.13" size="0.003" material="link" mass="0"/>
+        <geom name="mount" type="cylinder" fromto="0 0 0.12 0 0 0.1225" size="0.0075" material="qube" mass="0"/>
+        <body>
+            <joint name="theta" type="hinge" pos="0 0 0.13" axis="0 0 1" damping="[Dr]" frictionloss="0"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:60]: what new dp do you have in mind?

see below for my thought on renaming the params

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            <geom name="rotaryArm" type="cylinder" fromto="0 0 0.13 [Lr] 0 0.13" size="0.003" material="link" mass="[Mr]"/>
+            <geom name="rotaryArmExtension" type="cylinder" fromto="[Lr] 0 0.13 [0.0055+Lr] 0 0.13" size="0.003" material="link" mass="0"/>
+            <geom name="sensor" type="cylinder" fromto="-0.012 0 0.13 0.012 0 0.13" size="0.0065" material="link" mass="0"/>
+            <body>
+                <joint name="alpha" type="hinge" pos="[Lr] 0 0.13" axis="1 0 0" damping="[Dp]" frictionloss="0"/>
+                <geom name="pole" type="capsule" fromto="[Lr] 0 0.13 [Lr] 0 [0.13-Lp]" size="0.004" material="pole" mass="[Mp]"/>
+                <geom name="poleExtension" type="cylinder" fromto="[Lr] 0 0.13 [Lr] 0 0.135" size="0.004" material="pole" mass="0"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:67]: visual beauty only

File: Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        <geom name="upLink" type="cylinder" fromto="0 0 0.12 0 0 0.13" size="0.003" material="link" mass="0"/>
+        <geom name="mount" type="cylinder" fromto="0 0 0.12 0 0 0.1225" size="0.0075" material="qube" mass="0"/>
+        <body>
+            <joint name="theta" type="hinge" pos="0 0 0.13" axis="0 0 1" damping="[Dr]" frictionloss="0"/>
+            <geom name="rotaryArm" type="cylinder" fromto="0 0 0.13 [Lr] 0 0.13" size="0.003" material="link" mass="[Mr]"/>
+            <geom name="rotaryArmExtension" type="cylinder" fromto="[Lr] 0 0.13 [0.0055+Lr] 0 0.13" size="0.003" material="link" mass="0"/>
+            <geom name="sensor" type="cylinder" fromto="-0.012 0 0.13 0.012 0 0.13" size="0.0065" material="link" mass="0"/>
+            <body>
+                <joint name="alpha" type="hinge" pos="[Lr] 0 0.13" axis="1 0 0" damping="[Dp]" frictionloss="0"/>
+                <geom name="pole" type="capsule" fromto="[Lr] 0 0.13 [Lr] 0 [0.13-Lp]" size="0.004" material="pole" mass="[Mp]"/>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/assets/quanser_qube.xml:66]: Why? :D

File: Pyrado/pyrado/environments/mujoco/quanser_qube.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    goal_pos_init_sim_7dof,
+    init_qpos_des_4dof,
+    init_qpos_des_7dof,
+    wam_dgains_4dof,
+    wam_dgains_7dof,
+    wam_pgains_4dof,
+    wam_pgains_7dof,
+    wam_q_limits_lo_7dof,
+    wam_q_limits_up_7dof,
+)
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/quanser_qube.py:52]: not getting removed by isort for whatever reason

File: Pyrado/pyrado/environments/mujoco/quanser_qube.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+from pyrado.environments.mujoco.base import MujocoSimEnv
+from pyrado.environments.mujoco.wam_base import WAMSim
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/quanser_qube.py:54]: yes

File: Pyrado/pyrado/environments/mujoco/quanser_qube.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+            Rm=8.4,  # motor resistance [Ohm]
+            km=0.042,  # motor back-emf constant [V*s/rad]
+            Mr=0.095,  # rotary arm mass [kg]
+            Lr=0.085,  # rotary arm length [m]
+            Dr=5e-6,  # rotary arm viscous damping [N*m*s/rad], original: 0.0015, identified: 5e-6
+            Mp=0.024,  # pendulum link mass [kg]
+            Lp=0.129,  # pendulum link length [m]
+            Dp=1e-6,  # pendulum link viscous damping [N*m*s/rad], original: 0.0005, identified: 1e-6
+            V_thold_neg=0,  # min. voltage required to move the servo in negative the direction [V]
+            V_thold_pos=0,  # min. voltage required to move the servo in positive the direction [V]
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/quanser_qube.py:124]: Slightly longer names should never be an argument against readability. I strongly endorse the renaming!

File: Pyrado/pyrado/environments/mujoco/quanser_qube.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        )
+
+    def _mujoco_step(self, act: np.ndarray) -> dict:
+        assert self.act_space.contains(act, verbose=True)
+
+        km = self.domain_param["km"]
+        Rm = self.domain_param["Rm"]
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/environments/mujoco/quanser_qube.py:131]: wasn't for speed but readability, imho the dict access clutters the equations

[the remark this answers is not in the document]
Comment (famura) [reply Pyrado/pyrado/environments/mujoco/quanser_qube.py:112]: but you also do it in line 151 & 152. please be consistent then...

File: Pyrado/pyrado/environments/mujoco/assets/furuta_pendulum.xml [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+ OR TECHNICAL UNIVERSITY OF DARMSTADT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+-->
+
+<mujoco model="inverted pendulum">
```
Comment (famura) [inline Pyrado/pyrado/environments/mujoco/assets/furuta_pendulum.xml:31]: Furuta

**Review by famura — CHANGES_REQUESTED** [review]
just some minor stuff left

**Review by famura — APPROVED** [review]
well done

**Pull request merged**